### ISSUE TITLE
Gutenberg: Support RTL styles

### ISSUE THREAD

**Issue by gwwar** [Automattic/wp-calypso#26910]
When we use a Right-To-Left language in Calypso, let's make sure our editor styles work as expected. We'll likely need to coordinate with the SDK squad when we reach this ticket. (ping @dmsnell or @ehg )

For local testing I believe we need to set the `rtl` feature flag to true.

Note that this issue is for the native integration in Calypso at /gutenberg

**Comment by dmsnell** [comment]
Adding some screenshots from existing core Gutenberg RTL behavior

<img width="1195" alt="screen shot 2018-10-23 at 2 02 20 pm" src="https://user-images.githubusercontent.com/5431237/47390882-48a1c080-d6cd-11e8-9058-6aac3990a977.png">

<img width="688" alt="screen shot 2018-10-23 at 2 02 27 pm" src="https://user-images.githubusercontent.com/5431237/47390889-4b041a80-d6cd-11e8-8539-1f2b415a9a5c.png">

<img width="453" alt="screen shot 2018-10-23 at 2 02 47 pm" src="https://user-images.githubusercontent.com/5431237/47390895-4dff0b00-d6cd-11e8-9145-840b5728b594.png">

<img width="905" alt="screen shot 2018-10-23 at 2 03 29 pm" src="https://user-images.githubusercontent.com/5431237/47390899-4fc8ce80-d6cd-11e8-925c-8d902721b2bc.png">

**Comment by thehenrybyrd** [comment]
I did a little testing here as well, to compare the state of Calypso in Gutenberg with that of current Calypso.
In my Calypso in Gutenberg testing, I started a new post at wpcalypso.wordpress.com/gutenberg/post. Later, I went there again, after marking the site, and then later the account interface, as Hebrew.

Currently, in Calypso, the post editor handles RTL pretty well _if the account interface is set_ to an RTL language. Setting the _site_ as RTL doesn't make the post editor in Calypso (or Gutenberg in Calypso) work for the language.

**Here's how it RTL languages in a non-RTL interface Calypso look:**
<img width="500" alt="screen shot 2018-10-18 at 5 23 26 pm" src="https://user-images.githubusercontent.com/15107387/47162686-1e32ba80-d2fd-11e8-99f0-dfbd21b8b40d.png">

**The content turns out fine when published:**
<img width="500" alt="screen shot 2018-10-18 at 5 23 20 pm" src="https://user-images.githubusercontent.com/15107387/47162754-40c4d380-d2fd-11e8-8005-0d2bca86f629.png">



**After changing the account interface to an RTL language (Hebrew in this case), everything flips.
Here's how it looks in Calypso:**

<img width="500" alt="screen shot 2018-10-18 at 5 25 57 pm" src="https://user-images.githubusercontent.com/15107387/47163015-cb0d3780-d2fd-11e8-9e07-47d53bbf3a87.png">

**Again, in Gutenberg on Calypso it looks great after the interface switch:**
(though, note that the interface itself isn't translated)
<img width="500" alt="screen shot 2018-10-18 at 5 28 47 pm" src="https://user-images.githubusercontent.com/15107387/47163126-09a2f200-d2fe-11e8-93dd-ef179e080be5.png">

**Here's the text I used (the rest of my Hebrew typing was gibberish)**

> این برخی از محتوای تست برای زبان چپ به راست است.
> זהו תוכן לבדיקת שפה משמאל לימין.
> This is some testing content for a left to right language.  The above scripts are Persian and Hebrew. Both have the "." on the wrong side.

Once the interface is set to Hebrew, typing in English is really confusing, since the cursor is on the left, and all punctuation initially appears on the left, moving to the proper place if you keep typing. 
**Here's a screenshot:**

<img width="500" alt="screen shot 2018-10-18 at 5 33 19 pm" src="https://user-images.githubusercontent.com/15107387/47163285-5a1a4f80-d2fe-11e8-9cac-2cb58fe9ca09.png">

However, when typing in Hebrew (what I typed is gibberish, I don’t know Hebrew), it feels natural - the punctuation _should_ be on the left, and the words appear where I expect - where the cursor is. In addition, punctuation shows up in-line instead of moving around. 

**Highlighting (selecting text via the keyboard) is acting a bit funny**, but I’d defer to someone used to typing in an RTL language for that.
When trying to highlight text by clicking shift + option + < or >,  nothing happens **if** the cursor is at the end of a line. If the cursor is in the middle of text already, navigating with arrow keys and highlighting with shift + arrow keys works as expected.
I'm using Mac OS's built-in Hebrew - Querty keyboard for this testing. In notepad on my mac, navigation and highlighting work as expected.

**Sidenote:** It would be _awesome_ if we could detect RTL languages and create a block that automatically handles them. Then, we could have LTR and RTL languages in the same post/page, which would be _wonderful_. ❤❤

**Comment by blowery** [comment]
>  It would be awesome if we could detect RTL languages and create a block that automatically handles them. Then, we could have LTR and RTL languages in the same post/page, which would be wonderful. ❤❤

There's some prior art there in the [AutoDirection component](https://github.com/Automattic/wp-calypso/tree/master/client/components/auto-direction). Might be a useful starting point.

**Comment by mmtr** [comment]
I tested how the UI looks in Hebrew and it seems that the current status of the integration handles very well the RTL languages. I haven't noticed any difference from how Gutenberg in Core displays the editor, apart from untranslated messages (which should be addressed in #28724) and the lack of this button which allows to enter LTR content in a paragraph block.

<img width="673" alt="screen shot 2018-11-22 at 13 40 09" src="https://user-images.githubusercontent.com/1233880/48908237-fdb0cf80-ee69-11e8-8e53-efffdc2d6d8e.png">

I'll investigate a bit further why this button is missing in the toolbar, but I wonder if we should close this issue and, if necessary, open a new one for adding the LTR button. /cc @Automattic/lannister 

**Screenshots**

| Gutenberg in Core | Gutenberg in Calypso |
|---|---|
| <img width="1277" alt="screen shot 2018-11-22 at 13 36 47" src="https://user-images.githubusercontent.com/1233880/48904866-1e742780-ee60-11e8-8b16-89be71a28ea7.png"> | <img width="1275" alt="screen shot 2018-11-22 at 13 37 00" src="https://user-images.githubusercontent.com/1233880/48904877-2764f900-ee60-11e8-9a17-9afe3ec6a164.png"> |
| <img width="1275" alt="screen shot 2018-11-22 at 13 37 22" src="https://user-images.githubusercontent.com/1233880/48904936-64c98680-ee60-11e8-9288-54232c5c4138.png"> | <img width="1280" alt="screen shot 2018-11-22 at 13 37 35" src="https://user-images.githubusercontent.com/1233880/48904946-714ddf00-ee60-11e8-9033-e4fcc19db350.png"> |
| <img width="627" alt="screen shot 2018-11-22 at 13 38 28" src="https://user-images.githubusercontent.com/1233880/48904990-90e50780-ee60-11e8-8e1e-ff59c8daf09a.png"> | <img width="640" alt="screen shot 2018-11-22 at 13 38 13" src="https://user-images.githubusercontent.com/1233880/48904969-80cd2800-ee60-11e8-9331-40d6e2b45064.png"> |
| <img width="299" alt="screen shot 2018-11-22 at 13 39 21" src="https://user-images.githubusercontent.com/1233880/48905052-bf62e280-ee60-11e8-8d64-e0c5f5f472ab.png"> | <img width="306" alt="screen shot 2018-11-22 at 13 39 04" src="https://user-images.githubusercontent.com/1233880/48905032-ad813f80-ee60-11e8-8803-7144e6a92958.png"> |
| <img width="267" alt="screen shot 2018-11-22 at 13 40 25" src="https://user-images.githubusercontent.com/1233880/48905082-d6a1d000-ee60-11e8-9979-6403b4acbac3.png"> | <img width="268" alt="screen shot 2018-11-22 at 13 40 34" src="https://user-images.githubusercontent.com/1233880/48905094-dd304780-ee60-11e8-95ed-500508a4afdb.png"> |
| <img width="394" alt="screen shot 2018-11-22 at 13 40 56" src="https://user-images.githubusercontent.com/1233880/48905110-ecaf9080-ee60-11e8-8bf7-890e72374327.png"> | <img width="404" alt="screen shot 2018-11-22 at 13 41 03" src="https://user-images.githubusercontent.com/1233880/48905125-f933e900-ee60-11e8-907b-ad3583ddd873.png"> |
| <img width="423" alt="screen shot 2018-11-22 at 13 47 51" src="https://user-images.githubusercontent.com/1233880/48905147-09e45f00-ee61-11e8-9813-92f586524967.png"> | <img width="403" alt="screen shot 2018-11-22 at 13 42 41" src="https://user-images.githubusercontent.com/1233880/48905158-11a40380-ee61-11e8-9005-3dc559a72f9b.png"> |
| <img width="267" alt="screen shot 2018-11-22 at 13 48 32" src="https://user-images.githubusercontent.com/1233880/48905174-208ab600-ee61-11e8-837e-863be0a13398.png"> | <img width="271" alt="screen shot 2018-11-22 at 13 49 26" src="https://user-images.githubusercontent.com/1233880/48905234-487a1980-ee61-11e8-815d-06f3b1b381cd.png"> |
| <img width="304" alt="screen shot 2018-11-22 at 13 48 42" src="https://user-images.githubusercontent.com/1233880/48905191-297b8780-ee61-11e8-8716-9af1b6998871.png"> | <img width="300" alt="screen shot 2018-11-22 at 13 49 33" src="https://user-images.githubusercontent.com/1233880/48905247-50d25480-ee61-11e8-9adb-586ba3508ea0.png"> |
| <img width="635" alt="screen shot 2018-11-22 at 13 48 53" src="https://user-images.githubusercontent.com/1233880/48905210-3304ef80-ee61-11e8-8c52-4d593bc5efb2.png"> | <img width="637" alt="screen shot 2018-11-22 at 13 49 14" src="https://user-images.githubusercontent.com/1233880/48905224-3f894800-ee61-11e8-942b-ba890b7f1a5e.png"> |
| <img width="282" alt="screen shot 2018-11-22 at 13 50 56" src="https://user-images.githubusercontent.com/1233880/48905299-78c1b800-ee61-11e8-9535-eadc09232413.png"> | <img width="279" alt="screen shot 2018-11-22 at 13 50 42" src="https://user-images.githubusercontent.com/1233880/48905303-7eb79900-ee61-11e8-943f-18ae9dfb1bd4.png"> |

**Comment by mmtr** [comment]
Yeah, I made the test on the native integration (the one using the `/gutenberg/post` URL). Somehow, it's already handling the RTL styles.

Maybe the `mini-css-extract-plugin-with-rtl` plugin is doing that automatically?

**Comment by gwwar** [comment]
Oh interesting, yeah maybe some Calypso base handling.

**Comment by mmtr** [comment]
Lack of the 'Left to right' button addressed in #28820.

**Comment by mmtr** [comment]
I wanted to know why the RTL styles were already handled, and I have found something interesting.

During the build process, Calypso creates a second CSS bundle, which is processed to be RTL, thanks to the Webpack RTL Plugin. This RTL bundle is the one Calypso loads when the user language is RTL.

This means that when Calypso generates the CSS file for the [`gutenberg-editor` section](https://github.com/Automattic/wp-calypso/blob/b039dfdf20d9a0ff083d71edd4328d7f6798fdf6/assets/stylesheets/sections/gutenberg-editor.scss), it automatically creates another CSS file (using RTLCSS under the hood) to be loaded when the language is RTL. 

That's why we don't need to import the `node_modules/@wordpress/*/build-style/style-rtl.css` files, since Calypso already generates the RTL styles.

However, this approach presents a problem. Gutenberg defines [some rules](https://github.com/WordPress/gutenberg/blob/bd897f6cb78e7625ad264c7a487bfa950118c85c/packages/editor/src/components/block-list/style.scss#L846) that shouldn't be auto-filpped based on direction using the [RTLCSS Control Directives](https://rtlcss.com/learn/usage-guide/control-directives/) (i.e. `/*rtl:ignore*/`). These comments are not present in the `node_modules/@wordpress/*/build-style/style.css` files (the ones that Calypso imports), so these rules are auto-flipped when Calypso generates the RTL styles, causing some issues.

I can think of 2 ways of solving this.

### Option 1: Keep RTLCSS Control Directives
We should open a PR in the Gutenberg repo proposing a change that marks all the RTLCSS Control Directives with `!` so [they are kept](https://cssnano.co/optimisations/discardcomments/) when Gutenberg [generates the build](https://github.com/WordPress/gutenberg/blob/bd897f6cb78e7625ad264c7a487bfa950118c85c/webpack.config.js#L152). 

@jasmussen do you see any problem with this option? I'm not sure why not all the comments are marked with `!`.

Once merged (and the new packages published), the comments will be present on the `node_modules/@wordpress/*/build-style/style.css` files and Calypso will generate the RTL styles as expected.

**Pros**
- We don't need to change the Calypso build process.
- Easier approach.

**Cons**
- Less maintainable. This will work as long as both Gutenberg and Calypso use RTLCSS, so they both understand the Control Directives.
- Calypso generates the RTL CSS files instead of using the ones already provided by Gutenberg.

### Option 2: Use the RTL CSS files provided by Gutenberg

We should implement a mechanism in Calypso which will allow us to pick RTL styles that are already built instead of building them.

Unfortunately, I have no idea about how we can achieve this. @Automattic/team-calypso, do you have any clue?

**Pros**
- More maintainable. We wouldn't rely on how Gutenberg build the RTL styles, as we would consume the built files directly.

**Cons**
- We probably would need to change the Calypso build process (which is something not trivial to me).
- More complex approach?

@Automattic/lannister, thoughts?

**Comment by Copons** [comment]
Looping in @yoavf.

Off the top of my head:

Option 1 could involve a tweak on the `calypso-i18n` module, to make it skip the `/*rtl:ignore*/` rules.
Option 2 sounds more complicated to me 🤔

**Comment by jsnajdr** [comment]
> These comments are not present in the node_modules/@wordpress/*/build-style/style.css files (the ones that Calypso imports), so these rules are auto-flipped when Calypso generates the RTL styles

@mmtr The style files in the `build-style` folders of all published `@wordpress/*` packages have all the `/* rtl:ignore */` comments preserved. Bundling these CSS files into Calypso stylesheets and creating RTL versions of them should work without any issues. Or am I missing something?

**Comment by mmtr** [comment]
> @mmtr The style files in the build-style folders of all published @wordpress/* packages have all the /* rtl:ignore */ comments preserved. Bundling these CSS files into Calypso stylesheets and creating RTL versions of them should work without any issues. Or am I missing something?

You're right @jsnajdr. I was missing [this specific rule](https://github.com/WordPress/gutenberg/blob/bd897f6cb78e7625ad264c7a487bfa950118c85c/packages/editor/src/components/block-list/style.scss#L846) from the RTL styles loaded by Calypso and I thought it was caused by the lack of the RTLCSS Control Directives in the `build-style` folder, but they are indeed there, with and without the `!` mark.

After our conversation on Slack (p1543243025411600-slack-lannister) we realized that the problem is that Calypso generates the RTL styles from the files in `public/sections`, which are already autoprefixed and minified. Because of that minification, the RTLCSS Control Directives without the `!` mark are not preserved.

The new webpack CSS build pipeline already solves this by generating the RTL styles before minifying the CSS files, so this can be solved by moving the `../../../node_modules/@wordpress` imports from the `sections/gutenberg-editor.scss` file to a ES6 import inside a JS file in the Gutenberg section.

However, this problem is also affecting production, since the RTL styles are generated from files that don't have the `rtl:*` comments. So, it would be worth it to also fix the `build-css:sections` npm script.

I'll take a look at both options.
